src/clients/SpokePoolClient.ts

```diff
@@ -660,16 +713,23 @@ export class SpokePoolClient extends BaseAbstractClient {
 
       const dataForQuoteTime = await this.batchComputeRealizedLpFeePct(depositEvents);
       for (const [index, event] of Array.from(depositEvents.entries())) {
-        const rawDeposit = spreadEventWithBlockNumber(event) as DepositWithBlock;
-
-        // Derive and append the destination token, LP fee, quote block number and block timestamp from the event.
-        // @dev Deposit events may _also_ include early deposits, in which case we did not retrieve a block.
-        const deposit: DepositWithBlock = {
-          ...rawDeposit,
-          realizedLpFeePct: dataForQuoteTime[index].realizedLpFeePct,
-          destinationToken: this.getDestinationTokenForDeposit(rawDeposit),
-          quoteBlockNumber: dataForQuoteTime[index].quoteBlock,
-        };
+        const rawDeposit = spreadEventWithBlockNumber(event);
+        let deposit: DepositWithBlock;
+
+        if (this.isV3DepositEvent(event)) {
+          deposit = { ...(rawDeposit as v3DepositWithBlock) };
```

Comment:
Out of scope of the v3 change, but I would really like to have a system where the spread event functions preserve type information. But this is a larger change and not specific to v3.

Reply:
Yeah I definitely agree on this. The current method forces us to add a lot of type assertions.

I would FWIW be in favour of just supporting each event type individually by unpacking via an Ethers interface, rather than the coercion that we currently do.

Reply:
matt thanks for the push on this, i think unpacking events individually would be useful here as well as a utility function that takes in an object, an expected type, and then unpacks it prop by prop